[wkxcass] iP

wkxcass opened the pull request:
### MoiMoi

> "Your conscious mind, like the computer screen, is a focusing tool, not a storage place." – David Allen ([source](https://earnworthy.com/david-allen-quote/))

MoiMoi helps you _focus on what matters_, freeing up your mind from the clutter of tasks.

MoiMoi is an **easy-for-all** task management solution. It's,
- text-based 📝
- user-friendly 🤝
- ~fast~ _SUPER_ FAST to use ⚡

MoiMoi's features include,
- [ ] Task management
- [ ] Deadline tracking (coming soon)
- [ ] Reminder notifications (coming soon)

And, the best part – it's for **FREE**! 🎉

Getting MoiMoi is as easy as ABC:
1. Ensure that `Java 17` is installed in your computer.
2. Download the `.jar` file of the latest release, from [here](https://github.com/wkxcass/ip/releases/).
3. Move the `.jar` file into an empty folder.
4. Open a command window in that folder (right-click on the folder, then select `Open in Terminal`).
5. Run the command `java -jar "moimoi.jar"`.

If you are a Java programmer, you can use MoiMoi's source code to practice Java too! Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

wkxcass committed:
MoiMoi.java: implement skeletal version

Let's rename Duke.java into MoiMoi.java, and implement a skeletal
version of the chatbot that simply greets the user and exits.

wkxcass committed:
MoiMoi.java: add echoing functionality

Let's add an echoing functionality to MoiMoi.java such that it echoes
commands entered by the user, and exits when the user types the command
"bye".

wkxcass committed:
MoiMoi.java: implement "add" and "list" functionalities

Let's,
* add a functionality for the chatbot to store whatever text entered by
the user
* add a functionality to display (in list-form) all inputs that have
been stored, when requested by the user

wkxcass committed:
Add mark/unmark functionalities

Let's,
* create a class named "Task" to represent tasks stored and tracked by
the chatbot
* add a functionality for the chatbot to mark tasks as done
* add a functionality for the chatbot to unmark completed tasks (ie.
mark tasks as undone)

The "Task" class encapsulates a String description, and the completion
status of a task.

wkxcass committed:
Add Parser.java

Let's add a class named Parser to streamline input parsing, by
separating the command and arguments. This modular approach enhances
maintainability, making future modifications (if any) to input handling
more manageable.

wkxcass committed:
Add support for tracking tasks of various types

Let's,
* create subclasses of Task, namely Todo, Event and Deadline,
leveraging inheritance to extend shared properties and behaviors
* add functionalities for the chatbot to appropriately handle (eg.
store, track etc.) tasks of the aforementioned types

wkxcass committed:
Semi-automate text UI testing

The input/output redirection technique is used to semi-automate the
testing of MoiMoi.

Let's,
* edit runtest.bat, substituting any instance of "Duke" into
"MoiMoi"
* add a file containing a series of user inputs, named input.txt
* add a file containing expected outputs corresponding to the
aforementioned inputs, named EXPECTED.TXT

MoiMoi's logo has been changed (in MoiMoi.java). It now consists of
mappable ASCII characters for encoding.

wkxcass committed:
Add exception classes

Let's create exception classes to facilitate error-handling.

The abstract class "MoiMoiException" represents all exceptions
specific to MoiMoi. Its subclasses, "InvalidCommandException",
"MissingArgumentException" and "InvalidArgumentException" address
specific errors that MoiMoi may encounter during execution.

wkxcass committed:
Handle basic errors

Let's,
* Edit MoiMoi.java (and Parser.java) such that it appropriately
handles errors, mainly due to incorrect user inputs
* include error-related scenarios, to be tested on MoiMoi, in
input.txt and EXPECTED.TXT

wkxcass committed:
MoiMoi.java: add delete functionality

MoiMoi supports adding tasks, but is unable to delete tasks
specified by the user.

Users may then find it hard to manage their task lists effectively,
especially when they need to remove tasks.

Let's,
* modify MoiMoi.java to support deleting tasks as specified by the
user
* include corresponding tests in input.txt and EXPECTED.TXT

This ensures flexibility in MoiMoi's task management operations,
aligning with user expectations for basic task management
functionalities.

wkxcass committed:
MoiMoi.java: remove indentation for 'case' branches

The 'case' branches in MoiMoi.java are indented.

This can lead to unnecessary nesting and doesn't align with the
module's coding standards.

Let's remove indentations for 'case' branches in MoiMoi.java.

This improves code readability and ensures compliance with the
module’s coding standards.

wkxcass committed:
Add storage-specific String method for task classes

Tasks have a "toString" method that returns a String representation
of the task's information.

However, tasks need a separate way to format their information as a
String for storage purposes, which is different from the format
used in the UI.

Let's,
* create a new field in Task named "typeIcon" to generalize task
types
* implement a new method that returns a String of the task’s
information formatted for storage

This separation ensures that the data stored in files is consistent
with the expected format during loading, while keeping the UI
representation clean and focused on display purposes.

wkxcass committed:
Add storage-related exception classes

MoiMoi lacks specific exception classes to handle IO errors and
corrupted data files during storage operations.

Handling storage-related errors will make the codebase more robust and
allow clearer error messaging.

Let's,
* create an exception class, named "StorageCorruptedException", to
represent errors related to corrupted data files (i.e., when the content
is not in its expected format)
* create an exception class, named "StorageIOException", to represent IO
errors that may occur during data file creation or saving

These exceptions encapsulate storage-related errors, which can be easily
handled and managed.

wkxcass committed:
MoiMoi.java: add storage functionality

MoiMoi is unable to persist tasks between sessions.

Persistent storage is an expected task management feature, as users need
to save tasks and load them in future sessions.

Let's,
* create a class named "Storage" to encapsulate the data file used
to store tasks, along with methods for loading and saving. If the data
file does not exist, a new one will automatically be created
* modify MoiMoi.java to utilize the Storage class for task persistence

By centralizing storage logic in a class, the codebase becomes cleaner
and easier to maintain. Users are enabled to save and load tasks between
program runs.

wkxcass committed:
Update .gitignore

.gitignore does not account for the data files generated by MoiMoi
during its operation.

These files should not be tracked in the repository to avoid clutter.

Let's add an entry to .gitignore to exclude these data files from being
tracked by Git.

This ensures a more focused version control.

wkxcass committed:
Merge branch 'branch-Level-7'

wkxcass committed:
Add date-time-related exception classes

MoiMoi lacks specific exception classes to handle date-time errors.

Handling date-time-related errors will make the codebase more robust and
allow clearer error messaging.

Let's,
* create an exception class, named "InvalidDateTimeException", to
represent errors related to incorrect formatting of date and time
* create an exception class, named "InvalidDateTimeRangeException", to
represent errors in which the starting time of an event is later than
its ending time
* rename "InvalidArgumentException" as "InvalidIndexException" to
specifically handle errors related to incorrect task indices

"InvalildDateTimeException" and "InvalildDateTimeRangeException"
encapsulate date-time-related errors, making them easier to handle and
manage.

wkxcass committed:
Store date-times as java.time.LocalDateTime

Date-times are stored as strings in "Deadline" and "Event" tasks.

MoiMoi is unable to perform date-time validation, which can lead to
errors and reduces the effectiveness of task management.

Let's,
* store date-time information as "java.time.LocalDateTime" objects in
both "Deadline" and "Event" tasks
* modify MoiMoi.java and Storage.java to handle date-time parsing and
formatting
* include date-time-related scenarios, to be tested on MoiMoi, in
input.txt and EXPECTED.TXT

Utilizing LocalDateTime provides robust handling of date-time
information in the chatbot, facilitating better task management.

wkxcass committed:
Generalize InvalidDateTimeException

The "InvalidDateTimeException" is thrown only for "LocalDateTime"
parsing errors.

The "LocalDate" class to be introduced later can also throw parsing
errors. These are also expected to be handled using the
"InvalidDateTimeException" class.

Let's,
* generalize "InvalidDateTimeException" to represent errors thrown by
either "LocalDateTime" and "LocalDate" parsing methods
* ensure that the exception returns appropriate error messages specific
to the type of date-time object being parsed

This implementation prepares MoiMoi to handle both "LocalDateTime" and
"LocalDate" parsing errors. Users are able to receive clear and relevant
error messages if such issues arise.

wkxcass committed:
Add filter-by-date functionality

MoiMoi currently lacks functionality that makes proper use of date-time
objects after they are validated.

Handling date-time objects meaningfully can enhance task management,
specifically by allowing users to filter tasks by specific dates.

Let's,
* create a method within tasks to check if they occur on a specific date
* add a command for MoiMoi that filters and prints deadlines/events
occurring on a specific date
* include date-filtering-related scenarios, to be tested on MoiMoi, in
input.txt and EXPECTED.TXT

This provides meaningful use of date and time, as users are able to
filter and view tasks based on specific dates.

wkxcass committed:
Merge branch 'branch-Level-8'

wkxcass committed:
Update EXPECTED.TXT

The output format in EXPECTED.TXT has an inconsistent layout pattern.

Ensuring consistency in test output layouts helps maintain
predictability in testing.

Let's add an extra line at the end of the output in EXPECTED.TXT.

This standardizes the output format, improving the clarity and
consistency of test results.

wkxcass committed:
Extract Ui and TaskList classes

MoiMoi handles user interactions and task management directly within its
main class.

This makes the program not easy to maintain and extend.

Let's,
* extract a "Ui" class to handle I/O interactions with the user
* extract a "TaskList" class to manage the task list and provide
operations to add/delete tasks
* modify MoiMoi.java and Storage.java to utilize the aforementioned
classes, within their operations

This makes the codebase more modular and maintainable.

wkxcass committed:
Add command classes

MoiMoi's operations (and their corresponding commands) are represented
as strings.

This approach limits extensibility and makes the codebase hard to
maintain.

Let's,
* create an abstract "Command" class to represent chatbot commands
* implement subclasses of "Command" for each specific operation,
encapsulating the execution logic

This improves the code structure and makes it easier to extend or modify
commands in the future.

wkxcass committed:
MoiMoi.java: utilize Parser class and Command objects

MoiMoi.java contains both input parsing logic and the logic for
executing all operations.

This makes the codebase difficult to maintain and extend.

Let's,
* refactor "Parser" methods to convert user input directly into
executable Command objects
* remove parsing and command execution logic from MoiMoi.java, while
delegating these responsibilities to both "Parser" and "Command" classes

This simplifies MoiMoi.java and improves maintainability with a better
separation of concerns.

wkxcass committed:
Adjust UI display formats

Subtle modifications in MoiMoi's running logic have led to slight
changes in the UI display formats.

Adjustments are necessary to ensure that the output layout remains
consistent with the desired format.

Let's,
* modify Ui.java to align the output format with the updated program
logic
* update EXPECTED.TXT to reflect the desired output

This ensures that the UI remains intuitive, clear and consistent.

wkxcass committed:
Merge branch 'branch-A-MoreOOP'

wkxcass committed:
Use enumerations for commands

During input parsing, commands are directly converted from strings into
"Command" objects.

This approach lacks structure and is prone to errors due to direct raw
string handling.

Let's,
* create an Enum named "CommandEnum" to represent all command types
supported by MoiMoi
* modify Parser.java and AddCommand.java to utilize command enumerations

This enhances the robustness and clarity of the codebase, by replacing
raw string handling with a more structured approach.

wkxcass committed:
Create moimoi.command package

Command classes are located on the same directory level as other classes
used by MoiMoi.

This unorganized and cluttered project structure makes the codebase
harder to maintain as the number of classes increases.

Let's create a package named "moimoi.command" and move all
command-related classes into this package.

This improves code maintainability by grouping related classes together.
The codebase also becomes cleaner and easier to navigate.

wkxcass committed:
Create moimoi.exception package

Exception classes are located on the same directory level as other
classes used by MoiMoi.

This unorganized and cluttered project structure makes the codebase
harder to maintain as the number of classes increases.

Let's create a package named "moimoi.exception" and move all
exception classes into this package.

This improves code maintainability by grouping related classes together.
The codebase also becomes cleaner and easier to navigate.

wkxcass committed:
Create moimoi.task package

Task classes are located on the same directory level as other classes
used by MoiMoi.

This unorganized and cluttered project structure makes the codebase
harder to maintain as the number of classes increases.

Let's,
* create and utilize an Enum named "TaskEnum" representing all task
types supported by MoiMoi, to enhance code robustness
* create a package named "moimoi.task" and move all task-related classes
into this package

This improves code maintainability by grouping related classes together.
The codebase also becomes cleaner and easier to navigate.

wkxcass committed:
Move general classes to moimoi package

General classes are not organized under any package.

This lack of structure makes the codebase hard to navigate and maintain
as it grows.

Let's,
* move general classes into the "moimoi" package
* edit runtest.bat to include .java files from all packages, ensuring
that they are compiled and available during testing

This ensures that all classes related to MoiMoi are under the "moimoi"
package, with general classes in the root and grouped classes in sub
packages, improving clarity and maintainability.

wkxcass committed:
Merge branch 'branch-A-Packages'

wkxcass committed:
Merge branch 'add-gradle-support'

wkxcass committed:
Update Gradle configuration for iP

The initial Gradle configuration in build.gradle includes settings and
identifiers related to Duke's project setup.

These references prevent Gradle-associated tasks from running on
MoiMoi.

Let's,
* update the "application.mainClass" reference to reflect the correct
main class
* update the "shadowJar.archiveBaseName" reference

This ensures that Gradle can run tasks and produce correct outputs
aligned with MoiMoi's project setup.

wkxcass committed:
Adjust UI display formats

Minor flaws in the UI display formats are found.

Adjustments are necessary to ensure that the output layout remains
consistent with the desired format.

Let's modify Ui.java to align the output layouts with their desired
formats.

This ensures that the UI remains intuitive, clear and consistent.

wkxcass committed:
Merge branch 'branch-A-Gradle'

wkxcass committed:
Add JUnit tests

MoiMoi only has I/O comparison tests.

There are no dedicated tests for individual methods, which makes it hard
to ensure their correctness.

Let's add appropriate JUnit tests to test the behavior of non-trivial
methods in MoiMoi.

This improves test coverage and allows verification of the correctness
of each method's logic, ensuring overall robustness of MoiMoi.

wkxcass committed:
Merge branch 'branch-A-JUnit'

wkxcass committed:
Add JavaDoc comments

The codebase lacks detailed documentation.

This makes it hard for future developers to understand and
maintain the code.

Let's add header comments to all non-trivial classes and methods.

This improves code readability and maintainability, by providing
clear explanations of non-trivial classes and methods.

wkxcass committed:
Align code with course's coding standards

The codebase has some discrepancies when compared to the course's
coding standards.

Without adhering to coding standards, code can be hard to maintain
and navigate.

Let's make necessary adjustments to the code to comply with the
course's coding standards.

This ensures code consistency, clarity and maintainability.

wkxcass committed:
Add find-by-keyword functionality

MoiMoi does not support searching for tasks by keyword.

This limits MoiMoi's usability, as keyword search is an expected
functionality for effective task management.

Let's,
* add a command class named "FindCommand" that allows MoiMoi to
filter and print tasks containing a specific keyword in their
descriptions
* modify relevant classes to implement the "find" operation, by
integrating the "FindCommand" class

This improves user experience by allowing efficient task searching
and management.

wkxcass committed:
Add JUnit tests for find-by-keyword functionality

The newly added find-by-keyword functionality is untested.

Testing is essential to ensure that the "find" feature works as
expected.

Let's add JUnit tests to verify non-trivial methods used throughout
the "find" process.

This helps identify any issues early and ensures that the "find"
feature performs accurately under different scenarios.

wkxcass committed:
Update files for I/O redirection testing

Files for I/O redirection testing do not include test cases for the
newly added find-by-keyword feature.

These files need to be updated to reflect the increments made.

Let's,
* update input.txt and EXPECTED.TXT to include test cases for the
"find" feature
* adjust the configuration in runtest.bat to correctly redirect I/O
during testing
* modify StorageIOException.java to align with expected output
formats

This helps to validate that the find-by-keyword functionality and
its related outputs are working as intended.

wkxcass committed:
Merge branch 'branch-A-JavaDoc'

wkxcass committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/moimoi/exception/InvalidDateTimeException.java
#	src/main/java/moimoi/task/Task.java

Resolution: ensure adherence to the course's coding standards

wkxcass committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/moimoi/Parser.java
#	src/main/java/moimoi/Ui.java
#	src/main/java/moimoi/task/Task.java

Resolution: ensure correct positioning of JavaDoc comments and
adherence to the course's coding standards

wkxcass committed:
Merge branch 'branch-Level-9'

# Conflicts:
# src/main/java/moimoi/Parser.java
# src/main/java/moimoi/Ui.java
# src/main/java/moimoi/task/Task.java

Resolution: ensure correct positioning of JavaDoc comments and
adherence to the course's coding standards

wkxcass committed:
Merge branch 'master' of https://github.com/wkxcass/ip

# Conflicts:
#	src/main/java/moimoi/Ui.java

Resolution: retain line to align with code logic

wkxcass committed:
Rename method in "TaskList" class

The "TaskList" class contains a method named "size()".

This does not adhere to the course's coding standards, which
requires using verbs for method names.

Let's rename "size()" to "getSize()" and update all references to
the method accordingly.

This enhances code readability and clarity.

wkxcass committed:
Make keyword matching case-insensitive

Keyword matching during "find" operations is case-sensitive.

A case-insensitive searching is more user-friendly, as users are not
expected to remember the exact case of keywords.

Let's,
* update the "find" operation to use case-insensitive keyword
matching
* update input.txt, EXPECTED.TXT and relevant JUnit tests, to
include test cases for the new format

This improves functionality for the "find" feature, making it more
intuitive for users of MoiMoi.

wkxcass committed:
Integrate Checkstyle to enforce coding standards

There is no automated enforcement of coding standards within the
project.

Adhering to a consistent coding style is essential for more
maintainable and readable code.

Let's,
* create and update Checkstyle configuration files accordingly
* modify code to resolve violations identified by Checkstyle

This improves code quality, by ensuring that the codebase adheres
to a consistent coding standard.

wkxcass committed:
Merge branch 'branch-A-CheckStyle'

wkxcass committed:
Add GUI to chatbot program

MoiMoi uses a Command Line Interface (CLI).

This needs to change to provide a more user-friendly experience.

Let's,
* add necessary classes and FXML files, using JavaFX for Graphical
User Interface (GUI) development
* update the codebase to incorporate and utilize the GUI classes
* adapt JUnit test files accordingly, as well as remove classes,
references and test files related to the outdated CLI
* repackage classes (mainly reorganizing GUI and utility
classes) for a clearer and more maintainable project structure

Implementing an intuitive GUI enhances MoiMoi's visual interactions
with its users.

wkxcass committed:
Merge branch 'branch-Level-10'

wkxcass committed:
Tweak GUI

MoiMoi's GUI design is functional but can be further refined.

A change is needed to improve aesthetics.

Let's,
* add .css files as style sheets
* tweak GUI configurations to achieve the desired layout and design

The enhancement in GUI improves MoiMoi's overall user experience.

wkxcass committed:
Merge branch 'branch-Level-10'

Teddayz commented on src/main/java/moimoi/gui/DialogBox.java:
```diff
@@ -0,0 +1,82 @@
+     * Flips the dialog box such that the ImageView is on the left and text on the right.
+     */
+    private void flip() {
+        ObservableList<Node> tmp = FXCollections.observableArrayList(this.getChildren());
```
Perhaps could be named as tempList instead to give yourself more information what the variable is :)

Teddayz commented on src/main/java/moimoi/gui/DialogBox.java:
```diff
@@ -0,0 +1,82 @@
+     * @return Dialog box with the image of and text from MoiMoi.
+     */
+    public static DialogBox getMoiMoiDialog(String text, Image img) {
+        var db = new DialogBox(text, img);
```
Likewise over here, can give a better name for yourself to know what it is. Can be var dialogBox

Teddayz commented on src/main/java/moimoi/gui/MainWindow.java:
```diff
@@ -0,0 +1,90 @@
+     */
+    @FXML
+    public void initialize() {
+        scrollPane.vvalueProperty().bind(dialogContainer.heightProperty());
```
vValueProperty would be better

wkxcass replied:
Thank you for noticing! However there's nothing much I can do to it, since `vvalueProperty` is the name of an instance method of the `ScrollPane` class 😆

Teddayz commented on src/main/java/moimoi/util/exception/InvalidDateTimeException.java:
```diff
@@ -0,0 +1,37 @@
+     */
+    @Override
+    public String getMessage() {
+        return super.getMessage() + "The " + this.type
```
Maybe using String.format will reduce the length of this statement so that its more readable

Teddayz commented on src/main/java/moimoi/util/task/Deadline.java:
```diff
@@ -0,0 +1,62 @@
+ */
+public class Deadline extends Task {
+
+    protected LocalDateTime by;
```
Perhaps the name could be better, for example maybe we can use deadline as the name so that we know what the variable is :)

wkxcass replied:
Thanks! I agree with your feedback👍 I'll tweak my code accordingly to improve variable readability.

Teddayz reviewed (commented):
Overall, LGTM, has good coding style practices. 
One observation is that Deadline Tasks, event tasks etc can be renamed to DeadlineTask, EventTask, ToDoTask etc as it can make it more understandable that it is inherited from Task. 👍 Overall, good job!

JunLongling commented on src/main/java/moimoi/util/task/Task.java:
```diff
@@ -0,0 +1,91 @@
+     * Marks the task as done.
+     */
+    public void mark() {
+        this.isDone = true;
```
Hey, I noticed that this.isDone is used here, but since the field isn't shadowed by a parameter, I'm wondering if it's necessary. Should it be simplified to `isDone = true;` to reduce noise? I saw a similar pattern in a few other places as well—perhaps it's worth reviewing those instances too. Let me know what you think!

wkxcass replied:
Thank you so much for the insightful feedback! I hadn’t considered that perspective before.
My original intent behind using `this` was to remind myself that the variables are class instance scoped. However, you do make a point about reducing unnecessary noise, as `this` isn’t needed in most cases throughout the code. I’ll give this further consideration—it's indeed something worth thinking about, though I’m not ready to commit to refactoring just yet~

JunLongling commented on src/test/java/moimoi/util/command/AddCommandTest.java:
```diff
@@ -0,0 +1,169 @@
+    private TaskList tasks = new TaskList(new ArrayList<Task>());
+
+    @Test
+    public void testisExit() {
```
Would `testIsExit` be the correct naming standard here?

wkxcass replied:
Yes, seems like I overlooked this part 😆 Thanks for pointing it out!

JunLongling commented on src/main/java/moimoi/util/task/Deadline.java:
```diff
@@ -0,0 +1,62 @@
+     * @param date Date to be checked against the task's deadline.
+     * @return true if the task occurs on the specified date; false otherwise.
+     */
+    @Override
+    public boolean occursOn(LocalDate date) {
+        if (this.by.toLocalDate().isEqual(date)) {
+            return true;
+        } else {
+            return false;
+        }
+    }
```
Great job with the occursOn method! I see you’ve implemented the date comparison effectively using isEqual. While your approach works perfectly, you could consider simplifying the method by returning the result of the condition directly, like this: `return this.by.toLocalDate().isEqual(date);` This way, the logic remains the same but the code becomes more concise. Of course, there’s no right or wrong here—just a suggestion for clarity. What do you think?

wkxcass replied:
I like your suggestion! I'll tweak my code accordingly to make it more concise. Thanks for bringing up! 😄

JunLongling commented on src/main/java/moimoi/util/task/Task.java:
```diff
@@ -0,0 +1,91 @@
+     * Marks the task as undone (i.e., unmarks the task).
+     */
+    public void unmark() {
+        this.isDone = false;
```
Would it be better to simplify this to` isDone = false`; to reduce noise?

JunLongling reviewed (commented):
Overall, the coding standards are mostly adhered to (or at least, I can't detect many violations), good job! Keep up the great work! 👍

wkxcass committed:
Add assertions

The codebase does not include the use of the 'assert' feature.

This needs to change to document important assumptions and ensure
that they hold at various points in the code.

Let's add assertions to enforce these assumptions.

This improves code reliability by helping us catch potential bugs
earlier in the development process.

wkxcass committed:
Improve code quality

The code consists of areas that could be made more maintainable and
readable.

This needs to change to ensure that the code follows best practices
for maintainability and clarity.

Let's refactor the code where necessary.

By ensuring that the code aligns with established principles, this
improves overall code quality.

wkxcass committed:
Merge pull request #1 from wkxcass/branch-A-Assertions

Add assertions

wkxcass committed:
Merge branch 'master' into branch-A-CodeQuality

wkxcass committed:
Merge pull request #2 from wkxcass/branch-A-CodeQuality

Improve code quality

wkxcass committed:
Add support for managing fixed duration tasks

MoiMoi does not support fixed duration tasks that do not have a
specific start or end time.

This needs to change so that users are allowed to manage such tasks.

Let's add functionality to manage fixed duration (period) tasks.

This enhances MoiMoi's task management capabilities, making it more
versatile for various user needs.

wkxcass committed:
Merge branch 'branch-B-FixedDurationTasks'

wkxcass committed:
Set up continuous integration

The project lacks automated testing and integration.

This needs to change to ensure that the codebase is continuously
validated with each new commit or pull request.

Let's set up GitHub Actions for continuous integration (CI).

This improves code reliability by catching issues early in the
development process.

wkxcass committed:
Refactor JUnit tests

The JUnit tests could be better structured.

This needs to change to ensure that the tests are clear and
maintainable.

Let's refactor the JUnit tests.

This enhances test-code readability and maintainability.

wkxcass committed:
Refactor JUnit tests

The JUnit tests could be better structured.

This needs to change to ensure that the tests are clear and
maintainable.

Let's refactor the JUnit tests.

This enhances test-code readability and maintainability.

wkxcass committed:
Improve GUI

MoiMoi's GUI displays both sides of the conversation in the same visual
format.

This needs to change to better reflect the asymmetric nature of
interactions between the user and the program.

Let's tweak the GUI design to distinguish between user and MoiMoi dialog
boxes.

This makes the user interface more visually intuitive, and improves user
experience.

wkxcass committed:
Merge branch 'branch-A-BetterGui'

wkxcass committed:
Create user guide

The project lacks a user guide.

This makes it hard for new users to understand how to use MoiMoi.

Let's,
* create a comprehensive user guide
* include a screenshot of the GUI for visual reference

By providing clear instructions on how to interact with the chatbot,
this improves the user onboarding experience.

wkxcass committed:
Merge branch 'branch-A-UserGuide'

wkxcass committed:
README.md: fix typo

README.md contains a minor typo that inaccurately describes MoiMoi's
behavior.

This needs to change to ensure that the user guide is consistent
with the chatbot's behavior.

Let's fix the typo.

This keeps the user guide clear and correct, providing a reliable
reference for users.

wkxcass committed:
Merge branch 'branch-A-UserGuide'

wkxcass committed:
README.md: update formatting
